colexec: add Closers to NewColOperatorResult

asubiotto opened the pull request:
Release justification: bug fixes and low-risk updates to new functionality.
External operators would not be Closed if not drained completely, which is not
the case with a downstream limit. The flow would clean up disk space/file
descriptors, but this commit makes it so that Close is called on these
operators in cases where the operator is not fully drained.

These are only external storage operators that need to be explicitly told to
close when not drained of input (e.g. due to a downstream limit). When creating
a component, it can now be added to result.ToClose, which will add it to the
first downstream outbox or materializer (there must be at least one in the
flow) to close either on graceful or non-graceful termination.

Release note (bug fix): resources including disk space and file descriptors were
only being cleaned up at the end of a query's execution. This now happens as
early as possible.

Closes #44652

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46203)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
nit: I'd probably omit the release note entirely, or at least clarify that this could occur only with vectorized engine in the last month or so.

> Reviewed 17 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M2dX7M6ARMHJOjx4-W2:bgdxnzr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/disk_spiller.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dXQ1O0UTRhhi_2Jwv:-M2dXQ1O0UTRhhi_2Jww:b-tkl9m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/disk_spiller.go#L257)):*
> ```Go
> // Close closes the diskSpillerBase's input.
> // TODO(asubiotto): Remove this method. It only exists so that we can call Close
> //  from some runTests subtests when not draining the input fully. The test
> ```

super nit: could make the indentation of TODO the same with the rest (and in other places)?
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dXqoa6zdxv0HkT1jJ:-M2dXqoa6zdxv0HkT1jK:bo1chxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/external_hash_joiner_test.go#L82)):*
> ```Go
> 						//  function and do substring matching on the test name to
> 						//  conditionally explicitly call Close() on the hash joiner
> 						//  (through result.ToClose) in cases where it is know the sorter
> ```

nit: `s/know/known/g`.
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dY0rR7QJmKk3qZFP2:-M2dY0rR7QJmKk3qZFP3:b-s9cjww) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/external_hash_joiner_test.go#L92)):*
> ```Go
> 						// TODO(asubiotto): Explicitly Close when testing.T is passed into
> 						//  this constructor and we do a substring match.
> 						require.Equal(t, 3, len(closers))
> ```

Two merge joiners should be in `closers` as well.

Also, recently updated `relativeRank` operators need to be added to closers.
___
*[pkg/sql/colflow/vectorized_flow_test.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dZDrU3g-XqifuKnBn:-M2dZDrU3g-XqifuKnBo:b-cpzlae) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colflow/vectorized_flow_test.go#L200)):*
> ```Go
> 			require.Len(t, sources, 1)
> 			require.Len(t, inboxToNumInputTypes[sources[0].(*colrpc.Inbox)], numInputTypesToOutbox)
> 			return colrpc.NewOutbox(allocator, op, typs, sources, nil)
> ```

nit: it'd be nice to have an inline comment.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dZOJp38xdohrsO_wL:-M2dZOJp38xdohrsO_wM:blg9pdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colflow/colrpc/outbox.go#L62)):*
> ```Go
> 	draining        uint32
> 	metadataSources []execinfrapb.MetadataSource
> 	// closers is a slice of Closers that need to be closed on termination.
> ```

super nit: probably `s/closed/Closed/g` to stay consistent with other similar comments.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Omitted the release note.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M2hiBkfC6Kl9WMiofXk:b-rjr3bb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/disk_spiller.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dXQ1O0UTRhhi_2Jwv:-M2hEfV57iPeB9v4Ocmy:b4tvqbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/disk_spiller.go#L257)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: could make the indentation of TODO the same with the rest (and in other places)?
</blockquote></details>

goland highlights the whole TODO this way and I've been writing them like this for a while. Do you feel strongly about this?
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dY0rR7QJmKk3qZFP2:-M2hFsF51_o4Lzj0MIQ0:b-914org) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/external_hash_joiner_test.go#L92)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Two merge joiners should be in `closers` as well.

Also, recently updated `relativeRank` operators need to be added to closers.
</blockquote></details>

It depends on what we define the semantics of `Close` to be. Currently, we close the whole operator tree. I would prefer if there were only one closer here (the external hash joiner), but the two sorters are created through `createDiskBackedSort` that needs to add an external sorter to the close slice anyway. however, the merge joiner is created internally by the external hash joiner. I personally think this is fine for now since `Close` is idempotent.

Added `relativeRank` to `ToClose`
___
*[pkg/sql/colflow/colrpc/outbox.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dZOJp38xdohrsO_wL:-M2hGtqyAZfVDLH3Enwj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colflow/colrpc/outbox.go#L62)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: probably `s/closed/Closed/g` to stay consistent with other similar comments.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 20 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M2i7fqSDRcnBryNcDxE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/disk_spiller.go, line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dXQ1O0UTRhhi_2Jwv:-M2i69EDErOlRPqm3q6c:b-8lv0kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/disk_spiller.go#L257)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

goland highlights the whole TODO this way and I've been writing them like this for a while. Do you feel strongly about this?
</blockquote></details>

Hm, I didn't realize the highlighting bit, but such indentation is slightly annoying to my eye. I guess we can continue writing TODOs in our preferred ways (you with highlighting and me with constant indentation).
___
*[pkg/sql/colexec/external_hash_joiner_test.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M2dY0rR7QJmKk3qZFP2:-M2i6xLj-ZzG2sTmUmNK:brxp9sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f0fd3d11756df10a7e937a41e7d7b0eb651b88/pkg/sql/colexec/external_hash_joiner_test.go#L92)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It depends on what we define the semantics of `Close` to be. Currently, we close the whole operator tree. I would prefer if there were only one closer here (the external hash joiner), but the two sorters are created through `createDiskBackedSort` that needs to add an external sorter to the close slice anyway. however, the merge joiner is created internally by the external hash joiner. I personally think this is fine for now since `Close` is idempotent.

Added `relativeRank` to `ToClose`
</blockquote></details>

Sounds good.


<!-- Sent from Reviewable.io -->

asubiotto commented:
I'm realizing that `IdempotentClose` may be called by a separate goroutine. We probably want protection against that when setting up the materializer and outbox to call close. I'll fix this.

asubiotto commented:
Done. Added a `closerHelper`. PTAL.

asubiotto commented:
Actually this solution is misguided (I haven't slept much), the internal state needs to be protected from concurrent use in `Next`, similar to `DrainMeta`. Will ping when ready for review.

asubiotto commented:
Not sure what to do with this PR, I've also been focused on higher-priority things. To really add this functionality, we need to either protect all state that is being touched in both `DrainMeta` and `Close` or ensure that `Close` only gets called from the same goroutine. The first seems like not the right way to go and I'm not sure how much work the second option is. Semaphore resources will be released and leftover files removed in any case, I'm not sure what happens to the file handles when removing though. Putting this PR aside for now.

asubiotto commented:
I went down the route of protecting against concurrent `IdempotentClose` and `Next` calls by adding a mutex to the external hash joiner, sorter, rank operator, and merge joiner. I think this is the least invasive change. The alternative was to change our planning code to make the parallel unordered synchronizer aware of which components to close in which of its input goroutines as well as making the hash router aware of closers. I'm still on the fence so I'd like to get your opinion on this.

RFAL.

yuzefovich reviewed (commented):
Hm, I don't have any bright ideas on how to go around using mutex in `Next` and `IdempotentClose`.

I'm curious, though, whether the flow shutdown can become slower with this PR. Say we have an external hash joiner that is currently spilling its inputs (both are local `colBatchScan`s) to divide them into initial partitions. At the same time, `ConsumerDone` is called on the root materializer. IIUC, the external hash joiner will still return a single batch of joined rows, and only then the materializer will notice that its `ProcessorState` is no longer `Running` and will not request any more batches. Actually, this is happening regardless of this PR and might be the reason why on sqlsmith tests we often see "timed out but did not cancel execution" errors.

Hm, what about the ungraceful shutdown? How do rowflows get shutdown when we just ctrl-c the query?

> Reviewed 13 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M3HiEHf7vmYPIwP97uD:b-kt9bz7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/materializer.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3HhMNC5xEofA8QLrqC:-M3HhMNC5xEofA8QLrqD:b-zcj3yq) ([raw file](https://github.com/cockroachdb/cockroach/blob/08920b7bf0e0cccfc607f50836900bd78ad7c479/pkg/sql/colexec/materializer.go#L58)):*
> ```Go
> 	cancelFlow func() context.CancelFunc
> 
> 	// closers is a slice of Closers that should be Closed on termination.
> ```

nit: `s/Closers/IdempotentClosers/g`.
___
*[pkg/sql/colexec/operator.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3Hfd-PC3tAko25XN-o:-M3Hfd-PC3tAko25XN-p:b-7abiad) ([raw file](https://github.com/cockroachdb/cockroach/blob/08920b7bf0e0cccfc607f50836900bd78ad7c479/pkg/sql/colexec/operator.go#L172)):*
> ```Go
> 
> // closerHelper is a simple helper that helps Operators implement
> // IdempotentClose. If close returns true, resources may be released, if it
> ```

nit: `s/IdempotentClose/IdempotentCloser/g`.
___
*[pkg/sql/colexec/operator.go, line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3Hh_wCBsLzw6bReKsr:-M3Hh_wCBsLzw6bReKss:bw95fp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/08920b7bf0e0cccfc607f50836900bd78ad7c479/pkg/sql/colexec/operator.go#L176)):*
> ```Go
> // use.
> type closerHelper struct {
> 	closed bool
> ```

nit: maybe we should move the mutex into this struct?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
`ConsumerDone` is not expected to cancel anything and is only called from the same goroutine that `Materializer.Next` is called from (look at `Run` in `processorsbase.go`). `ConsumerDone` is only called in response when pushing rows to the receiver. If we return a zero-batch to the materializer, it will just drain its inputs as before in trailingMeta, *then* cancel the flow (will cancel any `colBatchScan` reading), *then* `Close` the hash joiner. The cancellation will unblock the hash joiner. So `Close`ing should not affect flow shutdown apart from the time it takes to close file descriptors/remove files if any.

When you cancel the given query, the context cancellation happens, which will unblock the col batch scan/hash joiner and be returned to the materializer which catches it as an error and drains, following the same procedure as before. The metadata may be meaningless in this case, but the flow is still cancelled *again*, probably pointlessly, and the operators `Close`d.

I think the important point here is that no method calls will unblock/cancel anything. The important part of query cancellation is context cancellation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M3LSBS68KceQj_fwUQz:bkpyu9j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/colexec/operator.go, line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3Hh_wCBsLzw6bReKsr:-M3LUtg54HZ85jWWU3zY:b-dz56bt) ([raw file](https://github.com/cockroachdb/cockroach/blob/08920b7bf0e0cccfc607f50836900bd78ad7c479/pkg/sql/colexec/operator.go#L176)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: maybe we should move the mutex into this struct?
</blockquote></details>

Not all structs that use a `closerHelper` need a mutex, I prefer to make it explicit.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Thanks, SGTM.

> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-:-M3MjkXm2VKPpBY_V9FF:bntszrz)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/mergejoiner.go, line 364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3MjpyaAPk2tBRl5ZeL:-M3MjpyaAPk2tBRl5ZeM:b-jygoua) ([raw file](https://github.com/cockroachdb/cockroach/blob/172caaf29ff4fb51b20523c84d95135926aa07d6/pkg/sql/colexec/mergejoiner.go#L364)):*
> ```Go
> 	// TODO(asubiotto): Explore calling IdempotentClose from the same goroutine as
> 	//  Next, which will simplify this model.
> 	mu syncutil.Mutex
> ```

There is a lint failure for copying this mutex. We might need to do something like:
```
mu struct {
  syncutil.Mutex
  _ util.NoCopy
}
```
___
*[pkg/sql/colexec/operator.go, line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3Hh_wCBsLzw6bReKsr:-M3MY9BC6asmOhdzULRJ:b-dxcbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/08920b7bf0e0cccfc607f50836900bd78ad7c479/pkg/sql/colexec/operator.go#L176)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Not all structs that use a `closerHelper` need a mutex, I prefer to make it explicit.
</blockquote></details>

Makes sense.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner.go, line 364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3MjpyaAPk2tBRl5ZeL:-M3MkVO63VY57XTYaknr:bkbtgr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/172caaf29ff4fb51b20523c84d95135926aa07d6/pkg/sql/colexec/mergejoiner.go#L364)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

There is a lint failure for copying this mutex. We might need to do something like:
```
mu struct {
  syncutil.Mutex
  _ util.NoCopy
}
```
</blockquote></details>

I think `NoCopy` is used just to detect cases like this, I changed the merge joiner base to be a pointer rather than a struct. It makes sense not to copy the base struct, what do you think?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46203)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/mergejoiner.go, line 364 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46203#-M3MjpyaAPk2tBRl5ZeL:-M3Ml0DA9uqa0xOZUZ3e:b-1bg6b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/172caaf29ff4fb51b20523c84d95135926aa07d6/pkg/sql/colexec/mergejoiner.go#L364)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think `NoCopy` is used just to detect cases like this, I changed the merge joiner base to be a pointer rather than a struct. It makes sense not to copy the base struct, what do you think?
</blockquote></details>

Oh yeah, that makes sense. Probably `mergejoiner_test.go` needs to be updated as well.


<!-- Sent from Reviewable.io -->

Commit:
colexec: add Closers to NewColOperatorResult

Release justification: bug fixes and low-risk updates to new functionality.
External operators would not be Closed if not drained completely, which is not
the case with a downstream limit. The flow would clean up disk space/file
descriptors, but this commit makes it so that Close is called on these
operators in cases where the operator is not fully drained.

These are only external storage operators that need to be explicitly told to
close when not drained of input (e.g. due to a downstream limit). When creating
a component, it can now be added to result.ToClose, which will add it to the
first downstream outbox or materializer (there must be at least one in the
flow) to close either on graceful or non-graceful termination.

Release note: None (no observable impact)

asubiotto commented:
bors r=yuzefovich
